src/Microsoft.DotNet.Interactive.HttpRequest.Tests/ParserTests.cs

```diff
@@ -0,0 +1,645 @@
+        public void common_verbs_are_parsed_correctly(string line, string method)
+        {
+            var result = HttpRequestParser.Parse(line);
+
+            result.SyntaxTree.RootNode
+                  .ChildNodes.Should().ContainSingle<HttpRequestNode>().Which
+                  .MethodNode.Text.Should().Be(method);
+        }
+
+        [Theory]
+        [InlineData("https://example.com?hat&ost=foo")]
+        [InlineData("https://example.com?q=3081#blah-2%203")]
+        public void common_url_structures_are_parsed_correctly(string url)
+        {
+            var result = HttpRequestParser.Parse($"GET {url}");
+
+            result.SyntaxTree.RootNode
+                  .ChildNodes.Should().ContainSingle<HttpRequestNode>().Which
+                  .UrlNode.Text.Should().Be(url);
+        }
+
+        [Theory]
+        [InlineData(@"GET https://example.com", "GET")]
+        [InlineData(@"Get https://example.com", "Get")]
+        [InlineData(@"OPTIONS https://example.com", "OPTIONS")]
+        [InlineData(@"options https://example.com", "options")]
+        public void it_can_parse_verbs_regardless_of_their_casing(string line, string method)
+        {
+            var result = HttpRequestParser.Parse(line);
+
+            result.SyntaxTree.RootNode
+                  .ChildNodes.Should().ContainSingle<HttpRequestNode>().Which
+                  .MethodNode.Text.Should().Be(method);
+        }
+
+        [Fact]
+        public void http_version_is_parsed_correctly()
+        {
+            var result = HttpRequestParser.Parse("GET https://example.com HTTP/1.1");
+
+            result.SyntaxTree.RootNode
+                  .ChildNodes.Should().ContainSingle<HttpRequestNode>().Which
+                  .VersionNode.Text.Should().Be("HTTP/1.1");
+        }
+        [Fact]
+        public void diagnostic_object_is_reported_for_unrecognized_verb()
+        {
+            var result = HttpRequestParser.Parse("OOOOPS https://example.com");
+
+            result.GetDiagnostics()
+                .Should().ContainSingle().Which.Message.Should().Be("Unrecognized HTTP verb OOOOPS");
+        }
+
+        [Fact]
+        public void request_node_without_method_node_created_correctly()
+        {
+            var result = HttpRequestParser.Parse("https://example.com");
+
+            result.SyntaxTree.RootNode.ChildNodes.Should().ContainSingle<HttpRequestNode>().Which
+                .MethodNode.Should().BeNull();
+        }
+    }
+
+    public class Headers
+    {
+        [Fact]
+        public void headers_are_parsed_correctly()
+        {
+            var result = HttpRequestParser.Parse(
+                """
+        GET https://example.com HTTP/1.1
+        Accept: */*
+        Accept-Encoding : gzip, deflate, br
+        Accept-Language : en-US,en;q=0.9
+        ContentLength:7060
+        Cookie: expor=;HSD=Ak_1ZasdqwASDASD;SSID=SASASSDFsdfsdf213123;APISID=WRQWRQWRQWRcc123123;
+        """);
```

Comment:
minor nit: Indentation seems off for many multi-line string literals in the file now. This probably happened when the code was copy pasted into nested classes...